impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java

```diff
@@ -47,140 +59,134 @@ public class KafkaClusterTest {
+        // TODO: waiting for new versions support in ozangunalp repo https://github.com/ozangunalp/kafka-native/issues/21
+        // version("3.3.1"),
+        // version("3.2.1")
+        );
     }
 
-    @Test
-    public void kafkaClusterZookeeperMode() throws Exception {
-        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
-                .testInfo(testInfo)
-                .kraftMode(false)
-                .build())) {
-            cluster.start();
-            verifyRecordRoundTrip(1, cluster);
-        }
+    private static Stream<BrokerCluster> clusterSizes() {
+        return Stream.of(
+                brokerCluster(1),
+                brokerCluster(2));
     }
 
-    @Test
-    public void kafkaTwoNodeClusterKraftMode() throws Exception {
-        int brokersNum = 2;
-        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
-                .testInfo(testInfo)
-                .brokersNum(brokersNum)
-                .kraftMode(true)
-                .build())) {
-            assumeTrue(cluster instanceof TestcontainersKafkaCluster, "KAFKA-14287: kraft timing out on shutdown in multinode case");
-            cluster.start();
-            verifyRecordRoundTrip(brokersNum, cluster);
-        }
+    private final Object[] brokerClusters = clusterSizes().toArray();
+
+    static Stream<List<Annotation>> clusterSizesAndModes() {
+        return Stream.of(
+                List.of(brokerCluster(1), kraftCluster(1)),
+                List.of(brokerCluster(2), kraftCluster(1)),
+                List.of(brokerCluster(1), zooKeeperCluster()),
+                List.of(brokerCluster(2), zooKeeperCluster()));
     }
 
-    @Test
-    public void kafkaTwoNodeClusterZookeeperMode() throws Exception {
-        int brokersNum = 2;
-        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
-                .testInfo(testInfo)
-                .brokersNum(brokersNum)
-                .kraftMode(false)
-                .build())) {
-            cluster.start();
-            verifyRecordRoundTrip(brokersNum, cluster);
+    @BeforeEach
+    public void initializeIndexes() {
+        if (clusterSizesIndex == clusterSizes().count()) {
+            clusterSizesIndex = 0;
         }
     }
 
-    @Test
-    public void kafkaClusterKraftModeWithAuth() throws Exception {
-        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
-                .kraftMode(true)
-                .testInfo(testInfo)
-                .securityProtocol("SASL_PLAINTEXT")
-                .saslMechanism("PLAIN")
-                .user("guest", "guest")
-                .build())) {
-            cluster.start();
-            verifyRecordRoundTrip(1, cluster);
-        }
+    @TestTemplate
+    @ExtendWith(KafkaClusterExtension.class)
+    public void kafkaClusterTemplate(@ConstraintsMethodSource("clusterSizesAndModes") InVMKafkaCluster cluster)
```

Comment:
As a test name this doesn't give me any clues what it's trying to demonstrate. So when I run the tests I get a nice sea of green but no idea what it has proved.

![Screenshot 2023-01-17 at 11 20 05 AM](https://user-images.githubusercontent.com/2833578/212773675-c9e488d4-d82d-48cf-8c79-0263ab2964c2.png)

Reply:
I can rename the test to show the goal of what we are testing, but the config if each test is pretty clear for me:
* We are testing in VM
* We know the number of brokers of each test
* We know if we are using zookeeper or KRaft mode

Reply:
Yeah we can see what configuration the cluster is in but what are we behaviour are we asserting on the cluster? 

I view the tests as an executable specification and that most of their value comes from enhancing our understanding of the requirements. Just telling me what the configuration of a cluster is only part of the story.

Reply:
Reworked.